File: annoycode.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Changes Unicode symbols with other symbols that looks the same.

from utils import checkPyVer
checkPyVer()

import sys
from data import Data

def usage():
    print("usage: {} <input file> <output file>".format(sys.argv[0]))
    print("Substitutes symbols with ones that look similar.")

if __name__ == "__main__":
    if len(sys.argv) != 3:
        usage()
        exit(-1)

    inPath = sys.argv[1]
    inData = ""
    try:
        with open(inPath, mode = "r") as f:
            inData = f.read()
    except:
        print("Could not read from input file: {}".format(inPath))
        exit(-1)

    if len(inData) == 0:
        print("Input file is empty: {}".format(inPath))
        exit(-1)

    outPath = sys.argv[2]
    outFile = None
    try:
        outFile = open(outPath, mode = "w+")
    except:
        print("Could not open output file for writing: {}".format(outPath))
        exit(-1)

    data = Data()
    if not data.load() or not data.hasMatches():
        print("Use trainer.py to generate data for use.")
        exit(-1)

    (out, subs) = data.subMatches(inData)
    if subs == 0:
        print("Nothing substituted")
        exit(1)

    outFile.write(out)
    outFile.close()

    inEnc = inData.encode("utf-8")
    outEnc = out.encode("utf-8")

    inCnt = len(inEnc)
    outCnt = len(outEnc)
    incPerc = float(outCnt) / float(inCnt) * 100 - 100

    print("{} substitutions, {} -> {} bytes, {}{}%".\
          format(subs, inCnt, outCnt, "+" if incPerc > 0 else "", incPerc))

File: src/annoycode.cpp
#include <QFile>
#include <QDebug>
#include <QCoreApplication>
#include <QCommandLineParser>

#include "Data.h"
#include "Constants.h"

int main(int argc, char **argv) {
  QCoreApplication app(argc, argv);

  QCommandLineParser parser;
  parser.setApplicationDescription("Substitues Unicode symbols that look "
                                   "similar.");
  parser.addHelpOption();
  parser.addPositionalArgument("input file", "File to read values from.",
                               "<input file>");
  parser.addPositionalArgument("output file", "File to write results to.",
                               "<output file>");

  QCommandLineOption
    fileOpt(QStringList() << "f" << "file",
            QString("Use substitutions from <data file>. Will "
                    "use \"%1\" otherwise.").arg(Consts::defaultDataFile),
            "data file");
  parser.addOption(fileOpt);

  parser.process(app);

  QStringList posArgs = parser.positionalArguments();
  if (posArgs.size() != 2) {
    qCritical() << "You must specify input and output file!";
    return -1;
  }

  QFile inFile(posArgs[0]);
  if (!inFile.open(QIODevice::ReadOnly)) {
    qCritical() << "Could not read from input file:" << posArgs[0];
    return -1;
  }

  QByteArray inData = inFile.readAll();
  if (inData.isEmpty()) {
    qCritical() << "Input file is empty!";
    return -1;
  }

  QString inText = QString::fromUtf8(inData);

  QFile outFile(posArgs[1]);
  if (!outFile.open(QIODevice::WriteOnly)) {
    qCritical() << "Could not open output file for writing:" << posArgs[1];
    return -1;
  }

  QString path;
  if (parser.isSet(fileOpt)) {
    path = parser.value(fileOpt);
  }

  auto data = Data(path);
  data.load();

  int count;
  QString res = data.substitute(inText, count);

  QByteArray outData = res.toUtf8();
  outFile.write(outData);
  outFile.close();

  auto inCnt = inData.size(),
    outCnt = outData.size();
  auto incPerc = float(outCnt) / float(inCnt) * 100 - 100;

  QString msg =
    QString("%1 substitutions applied, %2 -> %3 bytes, %4%5%")
    .arg(count).arg(inCnt).arg(outCnt).arg(incPerc > 0 ? "+" : "")
    .arg(incPerc);
  qDebug() << qPrintable(msg);

  return 0;
}
